elfi/methods/diagnostics.py

```diff
@@ -0,0 +1,260 @@
+        # preserve the parameters (thetas) corresponding to the `closest' datasets.
+        thetas = {}
+        E_me = np.inf
+        for set_ss in sets_candidate_ss:
+            names_ss = [ss.__name__ for ss in set_ss]
+            thetas_ss = self._obtain_accepted_thetas(set_ss, n_sim, n_acc)
+            thetas[set_ss] = thetas_ss
+            E_ss = self._calc_entropy(thetas_ss, n_acc, k)
+            if E_ss < E_me:
+                E_me = E_ss
+                names_ss_me = names_ss
+                thetas_closest = thetas_ss[:n_closest]
+            logger.info('Combination %s shows the entropy of %f' % (names_ss, E_ss))
+        logger.info('\nThe minimum entropy of %f was found in %s.\n' % (E_me, names_ss_me))
+
+        # Find the summary-statistics combination with
+        # the minimum mean root sum of squared error (MRRSE).
+        MRSSE_min = np.inf
+        for set_ss in sets_candidate_ss:
+            names_ss = [ss.__name__ for ss in set_ss]
+            MRSSE_ss = self._calc_MRSSE(set_ss, thetas_closest, thetas[set_ss])
+            if MRSSE_ss < MRSSE_min:
+                MRSSE_min = MRSSE_ss
+                names_ss_MRSSE = names_ss
+                set_ss_2stage = set_ss
+            logger.info('Combination %s shows the MRSSE of %f' % (names_ss, MRSSE_ss))
+        logger.info('\nThe minimum MRSSE of %f was found in %s.' % (MRSSE_min, names_ss_MRSSE))
+        return set_ss_2stage
+
+    def _create_ss_combinations(self, max_cardinality):
+        """Create all combinations of the initialised summary statistics up till the maximum cardinality.
+
+        Parameters
+        ----------
+        max_cardinality : int
+            Cardinality limit of a summary-statistics combination.
+
+        Returns
+        -------
+        array_like
+            Combinations of summary statistics.
+
+        """
+        if max_cardinality > len(self.list_ss):
+            max_cardinality = len(self.list_ss)
+
+        combinations_ss = []
+        for i in range(max_cardinality):
+            for combination in combinations(self.list_ss, i + 1):
+                combinations_ss.append(combination)
+        return combinations_ss
+
+    def _obtain_accepted_thetas(self, set_ss, n_sim, n_acc):
+        """Obtain the parameters accepted by the ABC-rejection sampling.
+
+        The sampling is performed using the initialised simulator.
+
+        Parameters
+        ----------
+        set_ss : array_like
+            Combination of the summary statistics used in the rejection sampling.
+        n_sim : int
+            Number of the iterations of the rejection sampling.
+        n_acc : int
+            Number of the accepted parameters.
+
+        Returns
+        -------
+        array_like
+            Accepted parameters.
+
+        """
+        # Initialise the distance function.
+        list_ss = []
+        for ss in set_ss:
+            list_ss.append(elfi.Summary(ss, self.simulator))
```

Comment:
Probably good to define the ElfiModel first and then give it as an argument to the created nodes

Reply:
If so, elfi complains that self.simulator corresponds to a different model than newly initialised elfi.new_model(). To familiarise with the self.simulator's initialisation, I have opened a PR with the diagnostics notebook, or you could look at test_diagnostics.py.

Since for every assessed summary-statistics combination, we initialise a new `d = elfi.Distance(self.fn_distance, *list_ss)`, ElfiModel's relevant internal attributes are adjusted as needed for the diagnostics.

Reply:
I mean you say `m = self.simulator.model.copy()` and then use `list_ss.append(elfi.Summary(ss, self.simulator, model=m))`. Otherwise you may end up modifying the original model.